### PR TITLE
Use a b-tree as the index structure within layer files

### PR DESCRIPTION
**Pull request by hlinnaka**
This reduces memory consumption of the page server, because the fairly large hashmaps are no longer "slurped" into memory without any accounting, but are cached by the page cache instead.

Fixes https://github.com/zenithdb/zenith/issues/1356.

### CODE REVIEW

**Comment by hlinnaka** [comment]
This is on top of PR #1350. I originally thought we should do this only after that is pushed, but actually perhaps we should just get this in now, because this changes the on-disk format again.

**Comment by hlinnaka** [comment]
This PR changes the layer file format to not use the 'bookfile' crate anymore. I thought that would be necessary, to get more control of the 8kB blocks of the pages and to use the page cache. But in hindsight, we probably could use the page cache even with the bookfile crate, too, so it was perhaps a mistake to change that as part of this PR. Thoughts?

File: pageserver/src/layered_repository.rs [hunk 1]
```diff
@@ -1760,6 +1762,7 @@ impl LayeredTimeline {
         // Now that we have reshuffled the data to set of new delta layers, we can
         // delete the old ones
         for l in level0_deltas {
+            println!("DELETINGX: {}", l.filename().display());
```
Comment (knizhnik) [inline pageserver/src/layered_repository.rs:1765]: Forget to comment debug print?

File: pageserver/src/layered_repository/delta_layer.rs [hunk 2]
```diff
@@ -534,61 +566,68 @@ impl DeltaLayerWriter {
...
+        let len = utils::write_blob(&mut self.bufwriter, &Value::ser(&val)?)?;
+
+        self.end_offset += len;
+
+        let will_init = match val {
+            Value::Image(_) => true,
+            Value::WalRecord(rec) => rec.will_init(),
+        };
+
+        let value = if will_init { off | 0x40_0000_0000 } else { off };
```
Comment (knizhnik) [inline pageserver/src/layered_repository/delta_layer.rs:582]: Ugh... Hardocoded constants

File: pageserver/src/page_cache.rs [hunk 3]
```diff
@@ -95,11 +91,15 @@ const MAX_USAGE_COUNT: u8 = 5;
 ///
 #[derive(Debug, PartialEq, Eq, Clone)]
 enum CacheKey {
-    MaterializedPage {
```
Comment (knizhnik) [inline pageserver/src/page_cache.rs:98]: Is is actually good idea to use the same cache both for ephemeral and layer files. May be maintain twoseparate caches, with the same implementation parameterized by key type?

File: pageserver/src/layered_repository/disk_btree.rs [hunk 4]
```diff
@@ -0,0 +1,3012 @@
...
+    /// generated by the 'test_random_updates' unit test. I extracted this while
+    /// trying to debug a failure in that test. The bug turned out to be
+    /// elsewhere, and I'm not sure if this is still useful, but keeping it for
+    /// now...  Maybe it's a useful data set to show the typical key-values used
+    /// by a delta layer, for evaluating how well the prefix compression works.
+    ///
+    #[test]
+    fn particular_data() -> anyhow::Result<()> {
+        #[rustfmt::skip]
+        static DATA:  [([u8; 26], u64); 2000] = [
```
Comment (LizardWizzard) [inline pageserver/src/layered_repository/disk_btree.rs:980]: If we keep this test data set I suggest to split it out to a separate file to increase readability. There is a handy macro [include_str!](https://doc.rust-lang.org/std/macro.include_str.html) that can help with that. Included string can have rows with two sort of columns separated by a space for example.

File: pageserver/src/layered_repository/disk_btree.rs [hunk 5]
```diff
@@ -0,0 +1,3012 @@
...
+    /// Last key that was appended to the tree. Used to sanity check that append
+    /// is called in increasing key order.
+    last_key: Vec<u8>,
+}
+
+impl<W> DiskBtreeBuilder<W>
+where
+    W: DiskBlockWriter,
+{
+    pub fn new(key_len: u8, writer: W) -> Self {
```
Comment (LizardWizzard) [inline pageserver/src/layered_repository/disk_btree.rs:451]: key_len can be turned into a const generic parameter, which should give more room for compiler optimizations and may help to remove assertions

**Comment by hlinnaka** [comment]
> The main concern I have at this moment is that there is soooo many new code, that it is almost not possible to review. But changing storage format several time is really bad idea, so +1 for merging it now.

Ok, I'll try to finish this up pretty soon then.

> My main technical concern is whether single page cache shared by all tenant will become bottleneck. Now almost all files access is performed through this cache. From one side, it is good to limit maximal amount of memory used by pageserver. But if we are going to serve thousands of tenants then we can easily get cache overflow (when all buffers are pinned and we can not find free one). Certainly we can block some thread in this case. But there is also risk of deadlock when some thread pins one buffer and need one more.

In principle, yeah. In practice it's highly improbable if the cache is large enough.

> Another bad thing with common cache is that read-once BLOB may flush out B-Tree internal pages.

Well, that's just a matter of what the cache eviction policy is.

> Concerning bookfile I am +1 for removing it. From my point of view it introduces more problems (troubles with buffered reader), than simplifying work with files. Certainly we can change it. But it seems to be non trivial. And in this case we will have to move bookfile to our repository, Not sure if it is possible. If we are going to support compression sometime, then it is one more reason to replace bookfile with something else.

Ok. I think I'll continue with the idea of removing bookfile then, but put a little more effort into commenting the exact file format and how the b-tree and the other stuff are laid out within the file.

File: pageserver/src/page_cache.rs (hunk 3 above)
Comment (hlinnaka) [reply pageserver/src/page_cache.rs:98]: I think it is better to use the same cache for as many things as possible. That way, the memory can be dynamically allocated to where it's needed the most. Maybe we should have different weighting in the eviction algorithm and evict non-dirty pages more eagerly than dirty ephemeral pages, or something like that. But would need more performance testing and analysis to figure out what the optimum is.

File: pageserver/src/layered_repository/disk_btree.rs (hunk 5 above)
Comment (hlinnaka) [reply pageserver/src/layered_repository/disk_btree.rs:441]: Oh cool, I didn't realize that that rust feature is stable already!

File: pageserver/src/layered_repository/disk_btree.rs (hunk 4 above)
Comment (hlinnaka) [reply pageserver/src/layered_repository/disk_btree.rs]: I moved it to a separate .rs file, without `include_str!`.

File: pageserver/src/layered_repository/disk_btree.rs (hunk 5 above)
Comment (hlinnaka) [reply pageserver/src/layered_repository/disk_btree.rs]: Done!

File: pageserver/src/repository.rs [hunk 6]
```diff
@@ -72,6 +84,16 @@ impl Key {
             field6: u32::from_be_bytes(b[14..18].try_into().unwrap()),
         }
     }
+
+    pub fn to_vec(&self) -> Vec<u8> {
+        let mut vec = vec![self.field1];
```
[the remark this answers is not in the document]
Comment (hlinnaka) [reply pageserver/src/repository.rs]: I replaced this with `fn write_to_byte_slice(&self, buf: &mut [u8])`.

[the remark this answers is not in the document]
Comment (hlinnaka) [reply pageserver/src/repository.rs]: > Comment on the Key struct references pgdatadir_mapping.rs but it no longer exists

Um, yes it does. And it seems unrelated to this piece of code, so I presume there is no issue and this comment was just a bogus?

[the remark this answers is not in the document]
Comment (LizardWizzard) [reply pageserver/src/repository.rs]: Oh, my bad, thats strange, my editor somehow failed to find it. II think it is pretty relevant here, Key is extensively used there: https://github.com/zenithdb/zenith/blob/heikki-kvstore/pageserver/src/pgdatadir_mapping.rs#L958 I'm just a bit confused with this fieldN style naming

**Comment by hlinnaka** [comment]
New version of this in PR #1459

**Pull request closed without merging**